soccer/src/soccer/strategy/agent/agent_action_client.cpp

```diff
@@ -121,4 +143,137 @@ void AgentActionClient::result_callback(const GoalHandleRobotMove::WrappedResult
+            send_unicast(communication_request.target_agents[0], request);
+        } else {
+            SPDLOG_WARN("BAD REQUEST NOT SENDING ANY COMMUNICATION");
+        }
+    }
+}
+
+void AgentActionClient::send_unicast(int robot_id, rj_msgs::msg::AgentRequest agent_request) {
+    auto request = std::make_shared<rj_msgs::srv::AgentCommunication::Request>();
+    request->agent_request = agent_request;
+
+    robot_communication_cli_[robot_id]->async_send_request(
+        request, [this, robot_id](
+                     const std::shared_future<rj_msgs::srv::AgentCommunication::Response::SharedPtr>
+                         response) { receive_response_callback(response, robot_id); });
+}
+
+void AgentActionClient::send_broadcast(rj_msgs::msg::AgentRequest agent_request) {
+    auto request = std::make_shared<rj_msgs::srv::AgentCommunication::Request>();
+    request->agent_request = agent_request;
+
+    for (size_t i = 0; i < kNumShells; i++) {
+        if (i != robot_id_ && this->world_state()->get_robot(true, i).visible) {
+            robot_communication_cli_[i]->async_send_request(
+                request,
+                [this,
+                 i](const std::shared_future<rj_msgs::srv::AgentCommunication::Response::SharedPtr>
+                        response) { receive_response_callback(response, i); });
+        }
+    }
+}
+
+void AgentActionClient::send_multicast(std::vector<u_int8_t> robot_ids,
+                                       rj_msgs::msg::AgentRequest agent_request) {
+    auto request = std::make_shared<rj_msgs::srv::AgentCommunication::Request>();
+    request->agent_request = agent_request;
+
+    for (int i : robot_ids) {
+        robot_communication_cli_[i]->async_send_request(
+            request,
+            [this,
+             i](const std::shared_future<rj_msgs::srv::AgentCommunication::Response::SharedPtr>
+                    response) { receive_response_callback(response, i); });
+    }
+}
+
+void AgentActionClient::receive_communication_callback(
+    const std::shared_ptr<rj_msgs::srv::AgentCommunication::Request>& request,
+    std::shared_ptr<rj_msgs::srv::AgentCommunication::Response>& response) {
+    // TODO: change this default to defense? or NOP?
+    if (current_position_ == nullptr) {
+        rj_msgs::msg::AgentResponse agent_response{};
+        rj_msgs::msg::Acknowledge acknowledgement{};
+        acknowledgement.acknowledged = true;
+        agent_response.acknowledge_response = {acknowledgement};
+        response->agent_response = agent_response;
+    } else {
+        // Convert agent request into AgentToPosCommRequest
+        rj_msgs::msg::AgentToPosCommRequest agent_request{};
+        agent_request.request = request->agent_request;
+
+        // Give the current position the request and receive the response to send back
+        rj_msgs::msg::PosToAgentCommResponse pos_to_agent_response =
+            current_position_->receive_communication_request(agent_request);
```

Comment:
Question: why not just have Position's `receive_communication_request()` take in an AgentRequest directly? What's the benefit of wrapping it in an AgentToPosCommRequest message?

I'm confused because the Position class has no ROS interfacing of its own (everything is handled for it by the AC) so it shouldn't have a choice in rejecting a comm req from the ActionClient, it should just process it and react. Like how Position doesn't have a choice when the AC updates its WorldState in AC's `world_state_callback()`.